lib/srv/heartbeatv2.go

```diff
@@ -73,20 +74,32 @@ func NewSSHServerHeartbeat(cfg SSHServerHeartbeatConfig) (*HeartbeatV2, error) {
 		return nil, trace.Wrap(err)
 	}
 
+	var metadataPtr atomic.Pointer[metadata.Metadata]
 	inner := &sshServerHeartbeatV2{
-		getServer: func(ctx context.Context) *types.ServerV2 {
-			server := cfg.GetServer()
-			metadata, err := metadata.Get(ctx)
-			if err == nil {
-				if metadata.CloudMetadata != nil {
-					server.SetCloudMetadata(metadata.CloudMetadata)
+		getMetadata: metadata.Get,
+		announcer:   cfg.Announcer,
+	}
+	inner.getServer = func(ctx context.Context) *types.ServerV2 {
+		server := cfg.GetServer()
+
+		meta := metadataPtr.Load()
+		if meta == nil {
+			go func() {
+				meta, err := inner.getMetadata(ctx)
+				if err != nil {
+					log.Warnf("Failed to get metadata: %v", err)
+				} else if meta != nil && meta.CloudMetadata != nil {
+					// Set the metadata immediately to give the heartbeat
+					// a chance to use it.
+					server.SetCloudMetadata(meta.CloudMetadata)
+					metadataPtr.CompareAndSwap(nil, meta)
 				}
-			} else {
-				log.Warnf("Failed to get metadata: %v", err)
-			}
-			return server
-		},
-		announcer: cfg.Announcer,
+			}()
+		} else if meta.CloudMetadata != nil {
+			server.SetCloudMetadata(meta.CloudMetadata)
```

Comment:
From comment thread above:

```suggestion
			// Server isn't cached between heartbeats, so set the metadata again.
			server.SetCloudMetadata(meta.CloudMetadata)
```